agent/engine/docker_task_engine.go

```diff
@@ -347,34 +355,50 @@ func (engine *DockerTaskEngine) wakeUpTaskQueueMonitor() {
 	}
 }
 
-func (engine *DockerTaskEngine) topTask() (*managedTask, error) {
+func (engine *DockerTaskEngine) validateIndex(index int) error {
+	waitingTaskQueueLength := len(engine.waitingTaskQueue)
+	// empty queue
+	if waitingTaskQueueLength == 0 {
+		return fmt.Errorf("no tasks in the waiting queue")
+	}
+	// invalid index
+	if index < 0 || index >= waitingTaskQueueLength {
+		return fmt.Errorf("invalid index: %v", index)
+	}
+	return nil
+}
+
+// getTaskByIndex returns the task from waitingTaskQueue at the given index
+func (engine *DockerTaskEngine) getTaskByIndex(index int) (*managedTask, error) {
 	engine.waitingTasksLock.Lock()
 	defer engine.waitingTasksLock.Unlock()
-	if len(engine.waitingTaskQueue) > 0 {
-		return engine.waitingTaskQueue[0], nil
+	err := engine.validateIndex(index)
+	if err != nil {
+		return nil, err
 	}
-	return nil, fmt.Errorf("no tasks in waiting queue")
+	return engine.waitingTaskQueue[index], nil
 }
 
 func (engine *DockerTaskEngine) enqueueTask(task *managedTask) {
 	engine.waitingTasksLock.Lock()
 	engine.waitingTaskQueue = append(engine.waitingTaskQueue, task)
 	engine.waitingTasksLock.Unlock()
-	logger.Debug("Enqueued task in Waiting Task Queue", logger.Fields{field.TaskARN: task.Arn})
+	logger.Debug("Enqueued task in the waiting task queue", logger.Fields{field.TaskARN: task.Arn})
 	engine.wakeUpTaskQueueMonitor()
 }
 
-func (engine *DockerTaskEngine) dequeueTask() (*managedTask, error) {
+// dequeueTaskByIndex dequeues a task from the waitingTaskQueue at the given index
+func (engine *DockerTaskEngine) dequeueTaskByIndex(index int) (*managedTask, error) {
 	engine.waitingTasksLock.Lock()
 	defer engine.waitingTasksLock.Unlock()
-	if len(engine.waitingTaskQueue) > 0 {
-		task := engine.waitingTaskQueue[0]
-		engine.waitingTaskQueue = engine.waitingTaskQueue[1:]
-		logger.Debug("Dequeued task from Waiting Task Queue", logger.Fields{field.TaskARN: task.Arn})
-		return task, nil
+	err := engine.validateIndex(index)
+	if err != nil {
+		return nil, err
 	}
-
-	return nil, fmt.Errorf("no tasks in waiting queue")
+	task := engine.waitingTaskQueue[index]
+	engine.waitingTaskQueue = append(engine.waitingTaskQueue[:index], engine.waitingTaskQueue[index+1:]...)
+	logger.Debug("Dequeued task from the waiting task queue", logger.Fields{field.TaskARN: task.Arn})
+	return task, nil
 }
 
 // monitorQueuedTasks starts as many tasks as possible based on FIFO order of waitingTaskQueue
```

Comment:
We should update the comment with the new behavior, as it will not be FIFO any more and try to start whichever task in the queue it can

Reply:
I'll update it. It is FIFO in terms of preference of evaluation but not blocking on tasks that unable to start.

Reply:
will remove the term "FIFO" to avoid confusion.

```
// monitorQueuedTasks starts as many tasks as possible based on FIFO order of waitingTaskQueue
```

Reply:
updated